### PR TITLE
Feature - Make the score type configurable through the result config object

### PR DESCRIPTION
**Pull request by smehringer**

### CODE REVIEW

File: include/seqan3/alignment/pairwise/edit_distance_fwd.hpp [hunk 1]
```diff
@@ -61,6 +61,16 @@ template <std::ranges::viewable_range database_t,
           typename word_t = uint_fast64_t>
 struct default_edit_distance_trait_type
 {
+    //!\brief The type of the alignment config.
+    using align_config_type = std::remove_reference_t<align_config_t>;
+
+private:
+    //!\brief The type score in the seqan3::alignment_result object (may be different from the internally used type).
```
Comment (eseiler) [inline include/seqan3/alignment/pairwise/edit_distance_fwd.hpp]:
```suggestion
    //!\brief The `score_type` of the seqan3::alignment_result object (may be different from the internally used type).
```

File: include/seqan3/alignment/pairwise/edit_distance_fwd.hpp [hunk 2]
```diff
@@ -91,9 +99,9 @@ struct default_edit_distance_trait_type
     static constexpr bool is_semi_global = is_semi_global_t::value;
     //!\brief Whether the alignment is a global alignment or not.
     static constexpr bool is_global = !is_semi_global;
-
     //!\brief Whether the alignment configuration indicates to compute and/or store the score.
-    static constexpr bool compute_score = align_config_type::template exists<align_cfg::result<with_score_type>>() ||
+    static constexpr bool compute_score = align_config_type::template exists<align_cfg::result<with_score_type,
+                                                                                               result_score_type>>() ||
```
Comment (marehr) [inline include/seqan3/alignment/pairwise/edit_distance_fwd.hpp]:
```suggestion
                                                                                               score_type>>() ||
```

File: include/seqan3/alignment/pairwise/edit_distance_fwd.hpp [hunk 3]
```diff
@@ -61,6 +61,16 @@ template <std::ranges::viewable_range database_t,
...
 {
+    //!\brief The type of the alignment config.
+    using align_config_type = std::remove_reference_t<align_config_t>;
+
+private:
+    //!\brief The type score in the seqan3::alignment_result object (may be different from the internally used type).
+    using result_score_type = std::conditional_t<align_config_type::template exists<align_cfg::result>(),
+                                                 typename std::remove_reference_t<
+                                                     decltype(get<align_cfg::result>(align_config_type{}))>::score_type,
+                                                 typename align_cfg::result<>::score_type>;
```
Comment (marehr) [inline include/seqan3/alignment/pairwise/edit_distance_fwd.hpp]: Can you do all this directly for the score_type definition below?

File: include/seqan3/alignment/pairwise/edit_distance_fwd.hpp [hunk 4]
```diff
@@ -61,6 +61,16 @@ template <std::ranges::viewable_range database_t,
...
+public:
     //!\brief The type of one machine word.
     using word_type = word_t;
     static_assert(std::is_unsigned_v<word_type>, "the word type of edit_distance_unbanded must be unsigned.");
```
Comment (marehr) [inline include/seqan3/alignment/pairwise/edit_distance_fwd.hpp]: `\/ \/ \/ \/ \/ \/ \/ \/ \/ \/ \/` score_type definition - below this `\/ \/ \/ \/ \/ \/ \/ \/ \/ \/ \/`

File: include/seqan3/alignment/pairwise/edit_distance_fwd.hpp [hunk 5]
```diff
@@ -70,8 +80,6 @@ struct default_edit_distance_trait_type
     using database_type = std::remove_reference_t<database_t>;
```
Comment (marehr) [inline include/seqan3/alignment/pairwise/edit_distance_fwd.hpp]: `/\ /\ /\ /\ /\ /\ /\ /\ /\ /\ /\` score_type definition - above this `/\ /\ /\ /\ /\ /\ /\ /\ /\ /\ /\`

File: include/seqan3/alignment/pairwise/edit_distance_fwd.hpp (hunk 3 above)
Comment (marehr) [reply include/seqan3/alignment/pairwise/edit_distance_fwd.hpp]: And we know that `align_cfg::result` exists, so this should be simplified. Add an `static_assert` that we expect this to be here. Maybe we need to fix some tests in edit distance to get this working.

File: include/seqan3/alignment/pairwise/align_result_selector.hpp [hunk 6]
```diff
@@ -117,7 +118,7 @@ struct align_result_selector
         }
         else
         {
-            return alignment_result_value_type<uint32_t, score_type>{};
+            return alignment_result_value_type<uint32_t, int32_t>{};
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/alignment/pairwise/align_result_selector.hpp]: Well the detail test, tests if this works with a config that does not contain an alignment result. All other tests run through with the change. So shall I change the test or leave the code?

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/alignment/pairwise/align_result_selector.hpp]: Change the test.

File: test/unit/alignment/pairwise/align_result_selector_test.cpp [hunk 7]
```diff
@@ -29,14 +29,6 @@ TEST(alignment_selector, align_result_selector_with_list)
...
-    { // test case I
-        auto cfg = align_cfg::edit;
-        using _t = alignment_result<typename detail::align_result_selector<seq1_t, seq2_t, decltype(cfg)>::type>;
-
-        EXPECT_TRUE((std::is_same_v<decltype(std::declval<_t>().id()), uint32_t>));
-        EXPECT_TRUE((std::is_same_v<decltype(std::declval<_t>().score()), int32_t>));
-    }
-
     { // test case II
         auto cfg = align_cfg::edit | align_cfg::result{with_score};
```
Comment (marehr) [inline test/unit/alignment/pairwise/align_result_selector_test.cpp]: can't you make this one case I and ... 

(or you have to change the numbering `// test case II` -> `// test case I`, `// test case III`)

File: test/unit/alignment/pairwise/align_result_selector_test.cpp [hunk 8]
```diff
@@ -63,6 +55,18 @@ TEST(alignment_selector, align_result_selector_with_list)
     }
 }
 
+TEST(alignment_selector, align_result_selector_using_score_type)
```
Comment (marehr) [inline test/unit/alignment/pairwise/align_result_selector_test.cpp]: ... this one case II?

Comment (smehringer) [reply test/unit/alignment/pairwise/align_result_selector_test.cpp]: I prefer single tests for single cases so I just removed the documentation from above. It just states the obvious anyway..

**Comment by marehr** [comment]
The snippets are failing: https://travis-ci.org/seqan/seqan3/jobs/618589436#L5802

**Comment by marehr** [comment]
The snippet failing is still a thing.

**Pull request merged**